argus/action_manager/windows.py

```diff
@@ -120,34 +121,85 @@ def get_installation_script(self):
 
+    def check_cbinit_installation(self):
+        """Check if Cloudbase-Init was installed successfully."""
+        LOG.info("Checking Cloudbase-Init installation.")
+
+        try:
+            python_dir = introspection.get_python_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
+        except exceptions.ArgusError as exc:
+            LOG.warning("Could not check Cloudbase-Init installation: %s", exc)
+            self.cbinit_cleanup()
+            return False
+
+        check_cmd = r'& "{}\python.exe" -c "import cloudbaseinit"'.format(
+            python_dir)
+        try:
+            self._client.run_remote_cmd(
+                cmd=check_cmd, command_type=util.POWERSHELL)
+        except exceptions.ArgusError as exc:
+            LOG.debug("Cloudbase-Init installation failed: %s", exc)
+            self.cbinit_cleanup()
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self):
+        """Clean-up Cloudbase-Init if it's installed."""
+        LOG.info("Cleaning up Cloudbase-Init.")
+
+        try:
+            cbinit_dir = introspection.get_cbinit_dir(
+                lambda cmd, command_type: self._client.run_command_with_retry(
+                    cmd, count=util.RETRY_COUNT,
+                    delay=util.RETRY_DELAY, command_type=command_type)[0]
+            )
+        except exceptions.ArgusError as exc:
+            LOG.warning("Could not find Cloudbase-Init directory: %s", exc)
+        else:
+            self.rmdir(ntpath.dirname(cbinit_dir))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Install Cloudbase-Init on the underlying instance."""
+        LOG.info("Trying to install Cloudbase-Init ...")
 
         installer = "CloudbaseInitSetup_{build}_{arch}.msi".format(
             build=self._conf.argus.build,
             arch=self._conf.argus.arch
         )
-        # TODO(cpopa): the service type is specific to each scenario,
-        # find a way to pass it
-        LOG.info("Run the downloaded installation script "
-                 "using the installer %r.", installer)
+
+        for _ in range(util.RETRY_COUNT):
+            for install_method in (self._run_installation_script,
+                                   self._deploy_using_scheduled_task):
+                try:
+                    install_method(installer)
+                except exceptions.ArgusError as exc:
+                    LOG.debug("Could not install Cloudbase-Init: %s", exc)
+                    self.cbinit_cleanup()
+                else:
+                    if self.check_cbinit_installation():
+                        return
```

Comment:
It's a good approach but it changes the method logic.
This method should try to install **Cloudbase-Init** using `_deploy_using_scheduled_task` method only in case of `_run_installation_script` method raises an exception. If `_run_installation_script` method runs normally but installation was invalid, main method should retry to install **Cloudbase-Init** using the same, `_run_installation_script`, method.